### PR TITLE
add graphs/ant_colony_optimization_algorithms.py

### PR DESCRIPTION
**Pull request by Clarkzzzzz**
### Describe your change:
A simple example uses the ant colony optimization algorithm to solve the classic TSP problem.

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

**Comment by Clarkzzzzz** [comment]
Hello, The file that failed the test is not my file. What should I do?

File: graphs/ant_colony_optimization_algorithms.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
+"""
+A simple example uses the ant colony optimization algorithm
+to solve the classic TSP problem
```
Comment (cclauss) [inline graphs/ant_colony_optimization_algorithms.py:3]: What is TSP and why does the reader need to guess?

File: graphs/ant_colony_optimization_algorithms.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+"""
+
+import copy
+import random
+
+
+def distance(city1: list[int], city2: list[int]) -> float:
+    """
+    Calculate the distance between two coordinate points
+    >>> distance([0,0],[3,4] )
```
Comment (cclauss) [inline graphs/ant_colony_optimization_algorithms.py:15]:
```suggestion
    >>> distance([0,0], [3,4] )
```

File: graphs/ant_colony_optimization_algorithms.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    5.0
+    """
+    return (((city1[0] - city2[0]) ** 2) + ((city1[1] - city2[1]) ** 2)) ** 0.5
+
+
+def pheromone_update(
+    pheromone: list[list],
```
Comment (cclauss) [inline graphs/ant_colony_optimization_algorithms.py:22]:
```suggestion
    pheromone: list[list[WHAT]],
```

File: graphs/ant_colony_optimization_algorithms.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    cities: dict[int, list[int]],
+    pheromone_evaporation: float,
+    ants_route: list[list[int]],
+    q: float,
```
Comment (cclauss) [inline graphs/ant_colony_optimization_algorithms.py:26]: What is `q` and why is the reader forced to guess?

File: graphs/ant_colony_optimization_algorithms.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    best_path: list,
```
Comment (cclauss) [inline graphs/ant_colony_optimization_algorithms.py:27]:
```suggestion
    best_path: list[WHAT],
```

File: graphs/ant_colony_optimization_algorithms.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    best_distance: float,
+) -> tuple[list[list], list, float]:
+    """
+    Update pheromones on the route and update the best route
```
Comment (cclauss) [inline graphs/ant_colony_optimization_algorithms.py:31]: doctests needed.

File: graphs/ant_colony_optimization_algorithms.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+def city_select(
+    pheromone: list[list[int]],
+    current_city: dict[int, list[int]],
+    unvisited_cities: dict[int, list[int]],
+    alpha: float,
+    beta: float,
+) -> tuple[dict[int, list[int]], dict[int, list[int]]]:
+    """
+    Choose the next city for ants
```
Comment (cclauss) [inline graphs/ant_colony_optimization_algorithms.py:62]: doctests needed.

File: graphs/ant_colony_optimization_algorithms.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+            cities,
+            PHEROMONE_EVAPORATION,
+            ants_route,
+            Q,
+            best_path,
+            best_distance,
+        )
+
+    print("Best Path:", best_path)
+    print("Best Distance:", best_distance)
```
Comment (cclauss) [inline graphs/ant_colony_optimization_algorithms.py:136]:
```suggestion
    print(f"{best_path = }")
    print(f"{best_distance = }")
```

**Review by cclauss — CHANGES_REQUESTED** [review]

**Review by cclauss — APPROVED** [review]
It would be better to detect if there are no `cities` or `ants_num < 1`or `iterations_num < 1` and raise ValueError.

**Pull request merged**